Comprehensive Type conversion unit tests.

benwulfe opened the pull request:
Introduces a unit test project for spanner ado.net
Adds the first set of unit tests which test storing and retrieving types.
Runtime changes (Type conversion changes):
  Fixes char conversions so it isnt (incorrectly) treated as a long and has its own handling.
  Added custom handling of decimal and ulong which cannot leverage the long codepath
  Fixes sbyte so that it leverages the typeof(long) codepath
  Ensures datetime+timestamp values can roundtrip as strings
  Fixed a bug with structs where the field member wasn't read correctly
  Allows conversion to bool of various types (convert allows it)
  Allows you to call GetFieldValue<List<X>>() where X is the list item type which recursively does conversion.
Also has 2 other typos/nonfunctional changes in PriorityList + ReliableStreamReader.
  ditto with GetFieldValue<X[]>  (typed arrays)
  ditto with Dictionary<, X> (the first generic argument must accept a string)

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/Google.Cloud.Spanner.Data.Tests.csproj:
```diff
@@ -0,0 +1,28 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
+1 -- done in the other PR

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs:
```diff
@@ -88,6 +90,34 @@ private SpannerDbType(TypeCode typeCode, RepeatedField<StructType.Types.Field> s
         }
 
+        /// <inheritdoc />
+        public override string ToString()
```
yes, for logs and debugger use.

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ValueConversion.cs:
```diff
@@ -76,21 +76,16 @@ public static object ConvertToClrType(this Value wireValue, V1.Type spannerType,
             }
 
-            if (targetClrType == typeof(char))
+            if (targetClrType == typeof(sbyte))
```
no char needs special handling and cannot leverage the long codepath.  it was removed on purpose.

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs:
```diff
@@ -137,14 +167,26 @@ internal V1.Type ToProtobufType()
+        /// You may use any type that implements IEnumerable as a source for the array.  The type of each
+        /// item is determined by <paramref name="elementType"/>.
+        /// When calling <see cref="SpannerDataReader.GetFieldValue(string)"/> the default type
+        /// is <see cref="ArrayList"/>. You may, however, specify any type that implements IList which
```
not particularly.  I did not want to force a generic on folks by default.  I can see about using List<object>

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ValueConversion.cs:
```diff
@@ -76,21 +76,16 @@ public static object ConvertToClrType(this Value wireValue, V1.Type spannerType,
+            if (targetClrType == typeof(sbyte))
             {
-                return Convert.ToChar((long) ConvertToClrTypeImpl(wireValue, spannerType, typeof(long)));
+                return Convert.ToSByte((long)ConvertToClrTypeImpl(wireValue, spannerType, typeof(long)));
```
you mean why use Convert.ToSByte when you can just cast long to sbyte?  I wanted consistency in explaining to developers what to expect (expect convert to 'yourtype' to be called).
I dont see a downside to using Convert...

Agree with the extra helper that does the cast.

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ValueConversion.cs:
```diff
@@ -180,24 +175,25 @@ public static Value ToValue(object value, SpannerDbType spannerDbType)
+                        // This allows us to convert back to a datetime reliably from the
+                        // resulting string (so roundtrip works properly if the developer uses
+                        // a string as a backing field for a datetime for whatever reason).
+                        return new Value { StringValue = XmlConvert.ToString((DateTime)value, XmlDateTimeSerializationMode.Utc) };
```
I'm using XmlConvert to parse RFC 3339 so I'm using it in both directions for clarity.
I would feel worse about doing say a parseexact and getting a slight variation in the format that XmlConvert might handle.

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ValueConversion.cs:
```diff
@@ -252,9 +248,37 @@ private static object ConvertToClrTypeImpl(Value wireValue, V1.Type spannerType,
+                }
+            }
+
+            if (targetClrType == typeof(char))
```
yes exactly.  to reduce the number of the cases here, earlier up I convert to others (eg long/double) if they can be done without loss of precision.

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ValueConversion.cs:
```diff
@@ -252,9 +248,37 @@ private static object ConvertToClrTypeImpl(Value wireValue, V1.Type spannerType,
+                    case Value.KindOneofCase.NullValue:
+                        return default(char);
+                    case Value.KindOneofCase.StringValue:
+                        if (spannerType.Code == TypeCode.Int64)
```
hopefully this makes more sense now.  int64 spanner types are encoded over the wire as strings.

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/TypeTests.cs:
```diff
@@ -0,0 +1,627 @@
+                {"ArrayField", SpannerDbType.ArrayOf(SpannerDbType.Int64)}
+            });
+
+        private static string S(string s) => $"\"{s}\"";
```
sure.  S was shorter.  I'm fine with Quote.

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/TypeTests.cs:
```diff
@@ -0,0 +1,627 @@
+
+        private static string S(string s) => $"\"{s}\"";
+
+        private static IEnumerable<string> GetStringsForArray()
```
i wanted some of the test cases to be a raw enumerable backed by a method to ensure that also worked.

benwulfe commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.UnitTests/TypeTests.cs:
```diff
@@ -0,0 +1,627 @@
+            yield return new object[] {"1.5", SpannerDbType.Float64, "1.5"};
+            yield return new object[] {null, SpannerDbType.Float64, "null"};
+
+            //Spanner type = Int64 tests.
```
true -- originally some were lossy in precision and I'll add some back and make them one way.

jskeet commented:
Thanks for all the extra info - I definitely need to take another pass with fresh eyes. That may or may not be today, unfortunately. My main point of concern is still in terms of cultures:

- Converting from CLR to Spanner for things like double to string, there's an argument for using the current culture, but it still makes me shudder
- Converting from Spanner to CLR, I'm pretty sure we should be using the invariant culture everywhere

As an aside, we should have tests for timestamps with nanosecond precision (which should lose information but still parse without error) - just thought about that earlier, and didn't want to forget...

benwulfe commented:
I'm thinking its much better to always use invariant culture for all of our conversions.
If the developer wants to convert something for display purposes, they can do this on their own using their own conversions after retrieving it in its native format.

I think this falls into the case of not creating a system that allows developers to easily fall into a trap.  And converting using current culture can easily lead them into a situation where on some machines values get stored differently in the database.  This is a far worse situation that forcing the developer to do culture aware conversions on their own.

The alternative would be to allow the developer to pass in an IFormatProvider to use for conversions.  We'd only be able to support this on our own methods (as opposed to some codepaths that ado.net controls).  I think we can do this if demand shows itself -- but the conversion layer is already fairly heavy and its worth getting feedback on it before adding more functionality at this point.

benwulfe commented:
Did a pass to make conversions invariant culture and added a roundtrip test using nanoseconds.
probably needs another round of looking over/discussion.

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ValueConversion.cs:
```diff
@@ -180,24 +176,25 @@ public static Value ToValue(object value, SpannerDbType spannerDbType)
+                    return new Value { StringValue = Convert.ToString(value, CultureInfo.InvariantCulture) };
                 case TypeCode.Int64:
-                    return new Value {StringValue = value.ToString()};
+                    return new Value {StringValue = Convert.ToInt64(value).ToString(CultureInfo.InvariantCulture) };
```
We don't know what type `value` is here, right? If it's a string, we should be using the invariant culture here, too.
(This is one thing I dislike about the `Convert` methods - they make it unclear whether we're performing string parsing, string formatting, or just a numeric/bitwise conversion such as `long` to `byte`.)

benwulfe replied:
will examine value to see if its a string incoming...

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ValueConversion.cs:
```diff
@@ -373,14 +441,15 @@ private static object ConvertToClrTypeImpl(Value wireValue, V1.Type spannerType,
                 }
                 //a bit of recursion here...
                 IDictionary dictionary = (IDictionary)Activator.CreateInstance(targetClrType);
+                var itemType = targetClrType.GetGenericArguments().Skip(1).FirstOrDefault() ?? typeof(object);
```
We could run into interesting problems here if what's passed in is an open generic type, but I think it's probably reasonable for that to fail.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PriorityList.cs:
```diff
@@ -19,7 +19,8 @@
 {
     /// <summary>
-    /// A simple sorted list that allows items to return equality for IComparable.CompareTo
+    /// A simple sorted list that sorts based on the result of CompareTo,
```
Is this change meant to be in this PR rather than the PriorityList one?

benwulfe replied:
i think I just noticed it when doing this PR.  If its not a big deal, would like to just leave it in here.  If its a big issue, I can create a separate PR.

jskeet commented on apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ReliableStreamReader.cs:
```diff
@@ -34,7 +34,6 @@ namespace Google.Cloud.Spanner.V1 {
     internal sealed class ReliableStreamReader : IDisposable
     {
         private AsyncServerStreamingCall<PartialResultSet> _currentCall;
-        private readonly object _sync = new object();
```
Not sure which PR this is meant to be in, but if there's no particularly logic place, I don't mind too much.

benwulfe replied:
was a random thing I spotted (unused field).

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/TypeTests.cs:
```diff
@@ -0,0 +1,643 @@
+            expected = StrippedString(expected);
+            actual = StrippedString(actual);
+            var unorderedMatchExpected = new HashSet<string>(expected.Split(','));
+            var unorderedMatchActual = new HashSet<string>(actual.Split(','));
```
Did parsing as a struct not work, or is that still TBD? (You could always add a TODO to do that later, to get this in sooner.)

benwulfe replied:
tbd -- working on it now.

benwulfe commented:
ok I squashed this down to a single commit, added the better conversion and a todo for culture specific tests.  Jon, you think we can get this in now?

jskeet commented on apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/TypeTests.cs:
```diff
@@ -0,0 +1,633 @@
+            yield return new object[] {"badjuju", SpannerDbType.Date, ""};
+        }
+
+        private string StrippedString(string s) => new StringBuilder(s)
```
Looks like you can now delete this. (Yay!)
